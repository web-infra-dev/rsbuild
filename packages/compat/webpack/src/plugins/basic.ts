import path from 'path';
import type { RsbuildPlugin } from '../types';
import { applyBasicPlugin } from '@rsbuild/shared';

/**
 * Provide some basic configs of webpack
 */
export const pluginBasic = (): RsbuildPlugin => ({
  name: 'plugin-basic',

  setup(api) {
    applyBasicPlugin(api);

    api.modifyWebpackChain(async (chain, { env }) => {
      // Disable webpack performance hints.
      // These logs are too complex
      chain.performance.hints(false);

      // This will be futureDefaults in webpack 6
      chain.module.parser.merge({
        javascript: {
          exportsPresence: 'error',
        },
      });

      if (env === 'development') {
        // Set correct path for source map
        // this helps VS Code break points working correctly in monorepo
        chain.output.devtoolModuleFilenameTemplate(
          (info: { absoluteResourcePath: string }) =>
            path.resolve(info.absoluteResourcePath).replace(/\\/g, '/'),
        );
      }
    });
  },
});
