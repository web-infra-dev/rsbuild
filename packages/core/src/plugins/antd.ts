import { getAntdMajorVersion } from '@modern-js/utils';
import type { RsbuildTarget, DefaultRsbuildPlugin } from '@rsbuild/shared';

export const pluginAntd = (): DefaultRsbuildPlugin => ({
  name: `plugin-antd`,

  setup(api) {
    api.modifyRsbuildConfig((builderConfig) => {
      builderConfig.source ??= {};

      if (
        builderConfig.source.transformImport === false ||
        builderConfig.source.transformImport?.some(
          (item) => item.libraryName === 'antd',
        )
      ) {
        return;
      }

      const antdMajorVersion = getAntdMajorVersion(api.context.rootPath);
      // antd >= v5 no longer need babel-plugin-import
      // see: https://ant.design/docs/react/migration-v5#remove-babel-plugin-import
      if (antdMajorVersion && antdMajorVersion < 5) {
        builderConfig.source ??= {};
        builderConfig.source.transformImport = [
          ...(builderConfig.source.transformImport || []),
          {
            libraryName: 'antd',
            libraryDirectory: useSSR(api.context.target) ? 'lib' : 'es',
            style: true,
          },
        ];
      }
    });
  },
});

export function useSSR(target: RsbuildTarget | RsbuildTarget[]) {
  return (Array.isArray(target) ? target : [target]).some((item) =>
    ['node', 'service-worker'].includes(item),
  );
}
