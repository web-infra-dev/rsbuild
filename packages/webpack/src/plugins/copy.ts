import { fs } from '@rsbuild/shared/fs-extra';
import type { CopyPluginOptions, BuilderPlugin } from '../types';

export const pluginCopy = (): BuilderPlugin => ({
  name: 'plugin-copy',

  setup(api) {
    api.modifyWebpackChain(async (chain, { CHAIN_ID }) => {
      const config = api.getNormalizedConfig();
      const { copy } = config.output;

      if (!copy) {
        return;
      }

      const { default: CopyPlugin } = await import(
        '../../compiled/copy-webpack-plugin'
      );

      const options: CopyPluginOptions = Array.isArray(copy)
        ? { patterns: copy }
        : copy;

      chain.plugin(CHAIN_ID.PLUGIN.COPY).use(CopyPlugin, [options]);
    });

    api.modifyWebpackConfig(async (config) => {
      const copyPlugin = config.plugins?.find(
        (item) => item?.constructor.name === 'CopyPlugin',
      ) as unknown as CopyPluginOptions;

      if (copyPlugin) {
        // If the pattern.context directory not exists, we should remove CopyPlugin.
        // Otherwise the CopyPlugin will cause the webpack to re-compile.
        const isContextNotExists = copyPlugin.patterns.every(
          (pattern) =>
            typeof pattern !== 'string' &&
            pattern.context &&
            !fs.existsSync(pattern.context),
        );
        if (isContextNotExists) {
          config.plugins = config.plugins?.filter(
            (item) => item?.constructor.name !== 'CopyPlugin',
          );
        }
      }
    });
  },
});
