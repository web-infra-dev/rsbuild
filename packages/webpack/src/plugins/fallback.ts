import { join } from 'path';
import {
  getDistPath,
  getFilename,
  resourceRuleFallback,
} from '@rsbuild/shared';
import type { RsbuildPlugin } from '../types';

export const pluginFallback = (): RsbuildPlugin => ({
  name: 'plugin-fallback',

  setup(api) {
    api.modifyWebpackChain((chain, { isProd }) => {
      const builderConfig = api.getNormalizedConfig();

      if (builderConfig.output.enableAssetFallback) {
        const distDir = getDistPath(builderConfig.output, 'media');
        const filename = getFilename(builderConfig.output, 'media', isProd);

        chain.output.merge({
          assetModuleFilename: join(distDir, filename),
        });
      }
    });

    api.modifyWebpackConfig((config) => {
      const builderConfig = api.getNormalizedConfig();

      if (!builderConfig.output.enableAssetFallback || !config.module) {
        return;
      }

      config.module.rules = resourceRuleFallback(config.module.rules);
    });
  },
});
