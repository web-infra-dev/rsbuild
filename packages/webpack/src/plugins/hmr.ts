import type { BuilderPlugin } from '../types';
import { isUsingHMR } from '@rsbuild/shared';

export const pluginHMR = (): BuilderPlugin => ({
  name: 'plugin-hmr',

  setup(api) {
    api.modifyWebpackChain((chain, utils) => {
      const config = api.getNormalizedConfig();

      if (!isUsingHMR(config, utils)) {
        return;
      }

      const { webpack, CHAIN_ID } = utils;
      chain.plugin(CHAIN_ID.PLUGIN.HMR).use(webpack.HotModuleReplacementPlugin);
    });
  },
});
