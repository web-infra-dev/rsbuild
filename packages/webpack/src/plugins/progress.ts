import { TARGET_ID_MAP } from '@rsbuild/shared';
import type { RsbuildPlugin } from '../types';

export const pluginProgress = (): RsbuildPlugin => ({
  name: 'plugin-progress',
  setup(api) {
    api.modifyWebpackChain(async (chain, { target, CHAIN_ID }) => {
      const config = api.getNormalizedConfig();
      const options = config.dev.progressBar;
      if (!options) {
        return;
      }

      const { ProgressPlugin } = await import(
        '../webpackPlugins/ProgressPlugin/ProgressPlugin'
      );
      chain.plugin(CHAIN_ID.PLUGIN.PROGRESS).use(ProgressPlugin, [
        {
          id: TARGET_ID_MAP[target],
          ...(options === true ? {} : options),
        },
      ]);
    });
  },
});
