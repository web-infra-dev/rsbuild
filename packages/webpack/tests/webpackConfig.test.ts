import { expect, describe, it } from 'vitest';
import { pluginBasic } from '@/plugins/basic';
import { createStubRsbuild } from './helper';
import { pluginBabel } from '@/plugins/babel';
import { pluginAntd } from '@rsbuild/core/plugins/antd';

describe('webpackConfig', () => {
  it('should allow tools.webpack to return config', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBasic()],
      builderConfig: {
        tools: {
          webpack(config) {
            return {
              ...config,
              devtool: 'eval',
            };
          },
        },
      },
    });

    const config = await rsbuild.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should allow tools.webpack to modify config object', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBasic()],
      builderConfig: {
        tools: {
          webpack(config) {
            config.devtool = 'eval';
          },
        },
      },
    });

    const config = await rsbuild.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should allow tools.webpack to be an object', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBasic()],
      builderConfig: {
        tools: {
          webpack: {
            devtool: 'eval',
          },
        },
      },
    });

    const config = await rsbuild.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should allow tools.webpack to be an array', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBasic()],
      builderConfig: {
        tools: {
          webpack: [
            {
              devtool: 'eval',
            },
            (config) => {
              config.devtool = 'source-map';
            },
          ],
        },
      },
    });

    const config = await rsbuild.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should provide mergeConfig util in tools.webpack function', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBasic()],
      builderConfig: {
        tools: {
          webpack: (config, { mergeConfig }) => {
            return mergeConfig(config, {
              devtool: 'eval',
            });
          },
        },
      },
    });

    const config = await rsbuild.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should allow to use tools.webpackChain to modify config', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBasic()],
      builderConfig: {
        tools: {
          webpackChain(chain) {
            chain.devtool('eval');
          },
        },
      },
    });

    const config = await rsbuild.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should allow tools.webpackChain to be an array', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBasic()],
      builderConfig: {
        tools: {
          webpackChain: [
            (chain) => {
              chain.devtool('eval');
            },
            (chain) => {
              chain.devtool('source-map');
            },
          ],
        },
      },
    });

    const config = await rsbuild.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should export HtmlWebpackPlugin instance', async () => {
    await createStubRsbuild({
      builderConfig: {
        tools: {
          webpack(config, utils) {
            expect(utils.HtmlWebpackPlugin.version).toEqual(5);
          },
        },
      },
    });
  });

  it('should allow to append and prepend plugins', async () => {
    const rsbuild = await createStubRsbuild({
      builderConfig: {
        tools: {
          webpack(config, utils) {
            utils.appendPlugins([new utils.webpack.DefinePlugin({ foo: '1' })]);
            utils.prependPlugins([
              new utils.webpack.DefinePlugin({ foo: '2' }),
            ]);
          },
        },
      },
    });

    const config = await rsbuild.unwrapWebpackConfig();
    expect(config.plugins).toMatchSnapshot();
  });

  it('should allow to remove plugins', async () => {
    const rsbuild = await createStubRsbuild({
      builderConfig: {
        tools: {
          webpack(config, utils) {
            utils.appendPlugins([new utils.webpack.DefinePlugin({ foo: '1' })]);
            utils.prependPlugins([
              new utils.webpack.DefinePlugin({ foo: '2' }),
            ]);
            utils.removePlugin('DefinePlugin');
          },
        },
      },
      plugins: [],
    });

    const config = await rsbuild.unwrapWebpackConfig();
    expect(config.plugins).toEqual([]);
  });

  it('should allow to add rules', async () => {
    const newRule = {
      test: /\.foo$/,
      loader: 'foo-loader',
    };

    const rsbuild = await createStubRsbuild({
      builderConfig: {
        tools: {
          webpack(config, utils) {
            utils.addRules(newRule);
          },
        },
      },
      plugins: [],
    });

    const config = await rsbuild.unwrapWebpackConfig();
    expect(config.module?.rules).toEqual([newRule]);
  });

  it('should set proper pluginImport option in Babel', async () => {
    // camelToDashComponentName
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBabel()],
      builderConfig: {
        source: {
          transformImport: [
            {
              libraryName: 'foo',
              camelToDashComponentName: true,
            },
          ],
        },
      },
    });
    const config = await rsbuild.unwrapWebpackConfig();

    const babelRules = config.module!.rules?.filter((item) => {
      // @ts-expect-error item has use
      return item?.use?.[0].loader.includes('babel-loader');
    });

    expect(babelRules).toMatchSnapshot();
  });

  it('should not set default pluginImport for Babel', async () => {
    // camelToDashComponentName
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBabel(), pluginAntd()],
    });
    const config = await rsbuild.unwrapWebpackConfig();

    const babelRules = config.module!.rules?.filter((item) => {
      // @ts-expect-error item has use
      return item?.use?.[0].loader.includes('babel-loader');
    });

    expect(babelRules).toMatchSnapshot();
  });

  it('should not have any pluginImport in Babel', async () => {
    // camelToDashComponentName
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBabel(), pluginAntd()],
      builderConfig: {
        source: {
          transformImport: false,
        },
      },
    });
    const config = await rsbuild.unwrapWebpackConfig();

    const babelRules = config.module!.rules?.filter((item) => {
      // @ts-expect-error item has use
      return item?.use?.[0].loader.includes('babel-loader');
    });

    expect(babelRules).toMatchSnapshot();
  });
});
