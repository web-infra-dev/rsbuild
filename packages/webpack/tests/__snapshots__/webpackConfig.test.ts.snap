// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`webpackConfig > should allow to append and prepend plugins 1`] = `
[
  DefinePlugin {
    "definitions": {
      "foo": "2",
    },
  },
  DefinePlugin {
    "definitions": {
      "process.env.ASSET_PREFIX": """",
      "process.env.NODE_ENV": ""test"",
    },
  },
  MiniCssExtractPlugin {
    "_sortedModulesCache": WeakMap {},
    "options": {
      "chunkFilename": "static/css/async/[name].css",
      "experimentalUseImportModule": undefined,
      "filename": "static/css/[name].css",
      "ignoreOrder": true,
      "runtime": true,
    },
    "runtimeOptions": {
      "attributes": undefined,
      "insert": undefined,
      "linkType": "text/css",
    },
  },
  HotModuleReplacementPlugin {
    "options": {},
  },
  ProgressPlugin {
    "compileTime": null,
    "dependenciesCount": 10000,
    "disableTsChecker": undefined,
    "handler": [Function],
    "hasCompileErrors": false,
    "id": "Client",
    "modulesCount": 5000,
    "name": "ProgressPlugin",
    "percentBy": null,
    "profile": false,
    "showActiveModules": false,
    "showDependencies": true,
    "showEntries": true,
    "showModules": true,
  },
  ReactRefreshPlugin {
    "options": {
      "exclude": [
        /node_modules/,
      ],
      "include": /\\\\\\.\\(\\[cm\\]js\\|\\[jt\\]sx\\?\\|flow\\)\\$/i,
      "overlay": false,
    },
  },
  DefinePlugin {
    "definitions": {
      "foo": "1",
    },
  },
]
`;

exports[`webpackConfig > should allow to use tools.webpackChain to modify config 1`] = `
{
  "context": "<ROOT>",
  "devtool": "eval",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should allow tools.webpack to be an array 1`] = `
{
  "context": "<ROOT>",
  "devtool": "source-map",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should allow tools.webpack to be an object 1`] = `
{
  "context": "<ROOT>",
  "devtool": "eval",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should allow tools.webpack to modify config object 1`] = `
{
  "context": "<ROOT>",
  "devtool": "eval",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should allow tools.webpack to return config 1`] = `
{
  "context": "<ROOT>",
  "devtool": "eval",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should allow tools.webpackChain to be an array 1`] = `
{
  "context": "<ROOT>",
  "devtool": "source-map",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should not have any pluginImport in Babel 1`] = `
[
  {
    "include": [
      {
        "and": [
          "<ROOT>",
          {
            "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
          },
        ],
      },
    ],
    "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
    "use": [
      {
        "loader": "<ROOT>/packages/shared/compiled/babel-loader",
        "options": {
          "babelrc": false,
          "compact": false,
          "configFile": false,
          "plugins": [
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
              {
                "version": "legacy",
              },
            ],
            "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
            "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
              {
                "proposal": "minimal",
              },
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
              {
                "useESModules": true,
                "version": "7.23.2",
              },
            ],
            "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
            [
              "<ROOT>/packages/webpack/compiled/babel-plugin-lodash",
              {},
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/babel-plugin-styled-components/lib/index.js",
              {
                "displayName": true,
                "pure": false,
                "ssr": false,
                "transpileTemplateLiterals": true,
              },
            ],
          ],
          "presets": [
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
              {
                "bugfixes": true,
                "corejs": {
                  "proposals": true,
                  "version": "3.32",
                },
                "exclude": [
                  "transform-typeof-symbol",
                ],
                "modules": "commonjs",
                "targets": [
                  "chrome >= 87",
                  "edge >= 88",
                  "firefox >= 78",
                  "safari >= 14",
                ],
                "useBuiltIns": "entry",
              },
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
              {
                "allExtensions": true,
                "allowDeclareFields": true,
                "allowNamespaces": true,
                "isTSX": true,
                "optimizeConstEnums": true,
              },
            ],
          ],
        },
      },
    ],
  },
  {
    "mimetype": {
      "or": [
        "text/javascript",
        "application/javascript",
      ],
    },
    "use": [
      {
        "loader": "<ROOT>/packages/shared/compiled/babel-loader",
        "options": {
          "babelrc": false,
          "compact": false,
          "configFile": false,
          "plugins": [
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
              {
                "version": "legacy",
              },
            ],
            "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
            "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
              {
                "proposal": "minimal",
              },
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
              {
                "useESModules": true,
                "version": "7.23.2",
              },
            ],
            "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
            [
              "<ROOT>/packages/webpack/compiled/babel-plugin-lodash",
              {},
            ],
          ],
          "presets": [
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
              {
                "bugfixes": true,
                "corejs": {
                  "proposals": true,
                  "version": "3.32",
                },
                "exclude": [
                  "transform-typeof-symbol",
                ],
                "modules": "commonjs",
                "targets": [
                  "chrome >= 87",
                  "edge >= 88",
                  "firefox >= 78",
                  "safari >= 14",
                ],
                "useBuiltIns": "entry",
              },
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
              {
                "allExtensions": true,
                "allowDeclareFields": true,
                "allowNamespaces": true,
                "isTSX": true,
                "optimizeConstEnums": true,
              },
            ],
          ],
        },
      },
    ],
  },
]
`;

exports[`webpackConfig > should not set default pluginImport for Babel 1`] = `
[
  {
    "include": [
      {
        "and": [
          "<ROOT>",
          {
            "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
          },
        ],
      },
    ],
    "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
    "use": [
      {
        "loader": "<ROOT>/packages/shared/compiled/babel-loader",
        "options": {
          "babelrc": false,
          "compact": false,
          "configFile": false,
          "plugins": [
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
              {
                "version": "legacy",
              },
            ],
            "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
            "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
              {
                "proposal": "minimal",
              },
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
              {
                "useESModules": true,
                "version": "7.23.2",
              },
            ],
            "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
            [
              "<ROOT>/packages/webpack/compiled/babel-plugin-lodash",
              {},
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/babel-plugin-styled-components/lib/index.js",
              {
                "displayName": true,
                "pure": false,
                "ssr": false,
                "transpileTemplateLiterals": true,
              },
            ],
          ],
          "presets": [
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
              {
                "bugfixes": true,
                "corejs": {
                  "proposals": true,
                  "version": "3.32",
                },
                "exclude": [
                  "transform-typeof-symbol",
                ],
                "modules": "commonjs",
                "targets": [
                  "chrome >= 87",
                  "edge >= 88",
                  "firefox >= 78",
                  "safari >= 14",
                ],
                "useBuiltIns": "entry",
              },
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
              {
                "allExtensions": true,
                "allowDeclareFields": true,
                "allowNamespaces": true,
                "isTSX": true,
                "optimizeConstEnums": true,
              },
            ],
          ],
        },
      },
    ],
  },
  {
    "mimetype": {
      "or": [
        "text/javascript",
        "application/javascript",
      ],
    },
    "use": [
      {
        "loader": "<ROOT>/packages/shared/compiled/babel-loader",
        "options": {
          "babelrc": false,
          "compact": false,
          "configFile": false,
          "plugins": [
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
              {
                "version": "legacy",
              },
            ],
            "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
            "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
              {
                "proposal": "minimal",
              },
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
              {
                "useESModules": true,
                "version": "7.23.2",
              },
            ],
            "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
            [
              "<ROOT>/packages/webpack/compiled/babel-plugin-lodash",
              {},
            ],
          ],
          "presets": [
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
              {
                "bugfixes": true,
                "corejs": {
                  "proposals": true,
                  "version": "3.32",
                },
                "exclude": [
                  "transform-typeof-symbol",
                ],
                "modules": "commonjs",
                "targets": [
                  "chrome >= 87",
                  "edge >= 88",
                  "firefox >= 78",
                  "safari >= 14",
                ],
                "useBuiltIns": "entry",
              },
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
              {
                "allExtensions": true,
                "allowDeclareFields": true,
                "allowNamespaces": true,
                "isTSX": true,
                "optimizeConstEnums": true,
              },
            ],
          ],
        },
      },
    ],
  },
]
`;

exports[`webpackConfig > should provide mergeConfig util in tools.webpack function 1`] = `
{
  "context": "<ROOT>",
  "devtool": "eval",
  "infrastructureLogging": {
    "level": "error",
  },
  "mode": "development",
  "module": {
    "parser": {
      "javascript": {
        "exportsPresence": "error",
      },
    },
  },
  "name": "Client",
  "performance": {
    "maxAssetSize": 3000000,
    "maxEntrypointSize": 3000000,
  },
}
`;

exports[`webpackConfig > should set proper pluginImport option in Babel 1`] = `
[
  {
    "include": [
      {
        "and": [
          "<ROOT>",
          {
            "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
          },
        ],
      },
    ],
    "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
    "use": [
      {
        "loader": "<ROOT>/packages/shared/compiled/babel-loader",
        "options": {
          "babelrc": false,
          "compact": false,
          "configFile": false,
          "plugins": [
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
              {
                "version": "legacy",
              },
            ],
            "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
            "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
              {
                "proposal": "minimal",
              },
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
              {
                "useESModules": true,
                "version": "7.23.2",
              },
            ],
            "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
            [
              "<ROOT>/node_modules/<PNPM_INNER>/babel-plugin-import/lib/index.js",
              {
                "camel2DashComponentName": true,
                "libraryName": "foo",
              },
              "foo",
            ],
            [
              "<ROOT>/packages/webpack/compiled/babel-plugin-lodash",
              {},
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/babel-plugin-styled-components/lib/index.js",
              {
                "displayName": true,
                "pure": false,
                "ssr": false,
                "transpileTemplateLiterals": true,
              },
            ],
          ],
          "presets": [
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
              {
                "bugfixes": true,
                "corejs": {
                  "proposals": true,
                  "version": "3.32",
                },
                "exclude": [
                  "transform-typeof-symbol",
                ],
                "modules": "commonjs",
                "targets": [
                  "chrome >= 87",
                  "edge >= 88",
                  "firefox >= 78",
                  "safari >= 14",
                ],
                "useBuiltIns": "entry",
              },
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
              {
                "allExtensions": true,
                "allowDeclareFields": true,
                "allowNamespaces": true,
                "isTSX": true,
                "optimizeConstEnums": true,
              },
            ],
          ],
        },
      },
    ],
  },
  {
    "mimetype": {
      "or": [
        "text/javascript",
        "application/javascript",
      ],
    },
    "use": [
      {
        "loader": "<ROOT>/packages/shared/compiled/babel-loader",
        "options": {
          "babelrc": false,
          "compact": false,
          "configFile": false,
          "plugins": [
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
              {
                "version": "legacy",
              },
            ],
            "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
            "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
              {
                "proposal": "minimal",
              },
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
              {
                "useESModules": true,
                "version": "7.23.2",
              },
            ],
            "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
            [
              "<ROOT>/node_modules/<PNPM_INNER>/babel-plugin-import/lib/index.js",
              {
                "camel2DashComponentName": true,
                "libraryName": "foo",
              },
              "foo",
            ],
            [
              "<ROOT>/packages/webpack/compiled/babel-plugin-lodash",
              {},
            ],
          ],
          "presets": [
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
              {
                "bugfixes": true,
                "corejs": {
                  "proposals": true,
                  "version": "3.32",
                },
                "exclude": [
                  "transform-typeof-symbol",
                ],
                "modules": "commonjs",
                "targets": [
                  "chrome >= 87",
                  "edge >= 88",
                  "firefox >= 78",
                  "safari >= 14",
                ],
                "useBuiltIns": "entry",
              },
            ],
            [
              "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
              {
                "allExtensions": true,
                "allowDeclareFields": true,
                "allowNamespaces": true,
                "isTSX": true,
                "optimizeConstEnums": true,
              },
            ],
          ],
        },
      },
    ],
  },
]
`;
